python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py

```diff
@@ -244,3 +244,175 @@ def test_avro_compression(rows, codec):
+                    "name": "o_date",
+                    "type": {
+                        "logicalType": "date",
+                        "type": "int",
+                    },
+                }
+            ],
+        }
+    )
+
+    dates = [
+        datetime.date(1970, 1, 1),
+        datetime.date(1970, 1, 2),
+        datetime.date(1981, 10, 25),
+        datetime.date(2012, 5, 18),
+        datetime.date(2019, 9, 3),
+    ]
+
+    days_from_epoch = [get_days_from_epoch(date) for date in dates]
+
+    records = [{"o_date": day} for day in days_from_epoch]
+
+    actual = cudf_from_avro_util(schema, records)
+
+    expected = cudf.DataFrame(
+        {"o_date": cudf.Series(dates, dtype="datetime64[s]")}
+    )
+
+    assert_eq(expected, actual)
+
+
+def get_total_microseconds_from_time(time: datetime.time) -> int:
+    return (
+        time.hour * 60 * 60 * 1000000
+        + time.minute * 60 * 1000000
+        + time.second * 1000000
+        + time.microsecond
+    )
+
+
+def get_total_milliseconds_from_time(time: datetime.time) -> int:
+    return (
+        time.hour * 60 * 60 * 1000
+        + time.minute * 60 * 1000
+        + time.second * 1000
+        + (time.microsecond // 1000)
+    )
+
+
+def test_can_parse_avro_time_logical_types():
+
+    schema = fastavro.parse_schema(
+        {
+            "type": "record",
+            "name": "test",
+            "fields": [
+                {
+                    "name": "f_time_millis",
+                    "type": {
+                        "logicalType": "time-millis",
+                        "type": "int",
+                    },
+                },
+                {
+                    "name": "f_time_micros",
+                    "type": {
+                        "logicalType": "time-micros",
+                        "type": "long",
+                    },
+                },
+            ],
+        }
+    )
+
+    time_tuples = [
+        # (hour, minute, second, microsecond)
+        (0, 0, 0, 0),
+        (0, 0, 0, 1),
+        (0, 0, 0, 10),
+        (0, 0, 0, 1000000 - 1),
+        (9, 30, 0, 0),
+        (23, 59, 59, 0),
+        (23, 59, 59, 1000000 - 1),
+    ]
+
+    times = [datetime.time(*time) for time in time_tuples]
+
+    micros = [get_total_microseconds_from_time(time) for time in times]
+    micros_as_ns = [micro * 1000 for micro in micros]
+
+    millis = [get_total_milliseconds_from_time(time) for time in times]
+    millis_as_ns = [milli * 1000000 for milli in millis]
+
+    records = []
+    for (milli, micro) in zip(millis, micros):
+        records.append({"f_time_millis": milli, "f_time_micros": micro})
+
+    actual = cudf_from_avro_util(schema, records)
+
+    expected = cudf.DataFrame(
+        {
+            "f_time_millis": cudf.Series(
+                millis_as_ns, dtype="timedelta64[ns]"
+            ),
+            "f_time_micros": cudf.Series(
+                micros_as_ns, dtype="timedelta64[ns]"
+            ),
+        }
+    )
```

Comment:
I'm not clear why the `cudf_from_avro_util` would return `timedelta64[ns]` dtypes here. I'd expect it to return `timedelta64[ms]` and `timedelta64[us]` respectively for the two columns based on the schema.

Reply:
Yeah that did seem odd, but as far as I could tell, it's an artifact of something downstream that handles the repr() of the dataframe and seems to normalize/convert everything to `timedelta64[ns]`?  Ditto for the date, you'd think that would be `datetime64[d]`, but it's `datetime64[s]`.

I'm just converting to the appropriate cudf type (i.e. `type_id::TIMESTAMP_DAYS`, `type_id::DURATION_MILLISECONDS`), and the *values* are correct, but, yeah, the dtype does seem off.

Reply:
In other words, I couldn't find any way to coerce the resulting `cudf.DataFrame` to have columns with dtypes *other* than `timestamp64[ns]` and `datetime64[s]`.

Reply:
That sounds roughly related to the fact that Pandas only supports `timedelta64[ns]`, but if we're constructing the `actual` with `cudf.read_avro` and the `expected` manually, we shouldn't be hitting Pandas along the way.

Reply:
Yeah, it's odd.  I can coerce `timestamp64[ms]` manually, e.g.:

```
[ins] In [11]: df = cudf.DataFrame({'time_millis': cudf.Series(millis, dtype='timedelta64[ms]')})

[ins] In [12]: df
Out[12]: 
             time_millis
0        0 days 00:00:00
1        0 days 00:00:00
2        0 days 00:00:00
3 0 days 00:00:00.999000
4        0 days 09:30:00
5        0 days 23:59:59
6 0 days 23:59:59.999000

[ins] In [13]: df.dtypes
Out[13]: 
time_millis    timedelta64[ms]
dtype: object
```

But you can see in my patch I'm not doing anything to ask for `timedelta64[ns]`, just storing the native decoded int directly:
```
      case type_timestamp_millis:
      case type_timestamp_micros:
      case type_local_timestamp_millis:
      case type_local_timestamp_micros:
      case type_time_millis:
      case type_time_micros: {
        // N.B. time-millis is stored as a 32-bit int, however, cudf expects an
        //      int64 for DURATION_MILLISECONDS.  From our perspective, the fact
        //      that time-millis comes from a 32-bit int is hidden from us by
        //      way of the zig-zag varint encoding, so we can safely treat them
        //      both as int64_t.  Everything else is 64-bit in both avro and
        //      cudf.
        int64_t v = avro_decode_zigzag_varint(cur, end);
        if (dataptr != nullptr && row < max_rows) { static_cast<int64_t*>(dataptr)[row] = v; }
      } break;
```

And in `reader_impl.cu` I'm just returning the cudf type, e.g. `type_id:DURATION_MILLISECONDS`:

```
type_id to_type_id(avro::schema_entry const* col)
{
  avro::type_kind_e kind;

  switch (col->logical_kind) {
    case avro::logicaltype_date:
    case avro::logicaltype_time_millis:
    case avro::logicaltype_time_micros:
      kind = static_cast<avro::type_kind_e>(col->logical_kind);
      break;
    case avro::logicaltype_not_set:
    default: kind = col->kind; break;
  }

  switch (kind) {
    case avro::type_boolean: return type_id::BOOL8;
    case avro::type_int: return type_id::INT32;
    case avro::type_long: return type_id::INT64;
    case avro::type_float: return type_id::FLOAT32;
    case avro::type_double: return type_id::FLOAT64;
    case avro::type_bytes:
    case avro::type_string: return type_id::STRING;
    case avro::type_date: return type_id::TIMESTAMP_DAYS;
    case avro::type_time_millis: return type_id::DURATION_MILLISECONDS;
    case avro::type_time_micros: return type_id::DURATION_MICROSECONDS;
    case avro::type_timestamp_millis: return type_id::TIMESTAMP_MILLISECONDS;
    case avro::type_timestamp_micros: return type_id::TIMESTAMP_MICROSECONDS;
    case avro::type_local_timestamp_millis: return type_id::TIMESTAMP_MILLISECONDS;
    case avro::type_local_timestamp_micros: return type_id::TIMESTAMP_MICROSECONDS;
    case avro::type_enum: return (!col->symbols.empty()) ? type_id::STRING : type_id::INT32;
    // There's no cuDF equivalent for the avro duration type, which is a fixed
    // 12 byte value which stores three little-endian unsigned 32-bit integers
    // representing months, days, and milliseconds, respectively.
    case avro::type_duration:
    default: return type_id::EMPTY;
  }
}
```

So something downstream in the repr() pipeline (or perhaps before) is converting into `timedelta64[ns]`, despite the original type being `timedelta64[ms]`.

Reply:
Ah, I see what's happening: `avro::type_time_millis` and `avro::type_time_micros`  represent a time of day with no specific calendar or date. The equivalent arrow types would be `TIME32` and `TIME64` respectively, but there's no libcudf equivalent type. The conversion functions we have rightfully throw on an unsupported Arrow data type which would be seen if trying to directly construct the cudf dataframe from the `times` list of `datetime.time` objects.

The `timedelta64[ns]` are coming up because Pandas only supports the nanosecond resolution for `timedelta64` dtypes as of now and the `assert_eq` function converts to Pandas DataFrames: https://github.com/rapidsai/cudf/blob/4e32bfe3aadf87716a91aed79048dec7dbe50e09/python/cudf/cudf/testing/_utils.py#L102-L105

Reply:
Ah!  That tracks.  Didn't realize there was some implicit pandas conversion happening in assert_eq().  The cudf df definitely has ms/us after all:
```
[ins] In [12]: df = cudf_from_avro_util(schema, records)

[ins] In [13]: df
Out[13]: 
           f_time_millis          f_time_micros
0        0 days 00:00:00        0 days 00:00:00
1        0 days 00:00:00 0 days 00:00:00.000001
2        0 days 00:00:00 0 days 00:00:00.000010
3 0 days 00:00:00.999000 0 days 00:00:00.999999
4        0 days 09:30:00        0 days 09:30:00
5        0 days 23:59:59        0 days 23:59:59
6 0 days 23:59:59.999000 0 days 23:59:59.999999

[ins] In [14]: df.dtypes
Out[14]: 
f_time_millis    timedelta64[ms]
f_time_micros    timedelta64[us]
dtype: object
```

I'll tweak the test a bit to make this clearer.

Reply:
Regardless, I don't think the mapping of avro types `time-millis` and `time-micros` have libcudf (or cudf Python) type equivalents, where they likely shouldn't be handled today. I think the mapping should look like:
```
  switch (kind) {
    case avro::type_boolean: return type_id::BOOL8;
    case avro::type_int: return type_id::INT32;
    case avro::type_long: return type_id::INT64;
    case avro::type_float: return type_id::FLOAT32;
    case avro::type_double: return type_id::FLOAT64;
    case avro::type_bytes:
    case avro::type_string: return type_id::STRING;
    case avro::type_date: return type_id::TIMESTAMP_DAYS;
    case avro::type_time_millis:
    case avro::type_time_micros:
    case avro::type_timestamp_millis: return type_id::TIMESTAMP_MILLISECONDS;
    case avro::type_timestamp_micros: return type_id::TIMESTAMP_MICROSECONDS;
    case avro::type_local_timestamp_millis: return type_id::TIMESTAMP_MILLISECONDS;
    case avro::type_local_timestamp_micros: return type_id::TIMESTAMP_MICROSECONDS;
    case avro::type_enum: return (!col->symbols.empty()) ? type_id::STRING : type_id::INT32;
    // There's no cuDF equivalent for the avro duration type, which is a fixed
    // 12 byte value which stores three little-endian unsigned 32-bit integers
    // representing months, days, and milliseconds, respectively.
    case avro::type_duration:
    default: return type_id::EMPTY;
  }
```

There's also `avro::type_decimal` which should map to the libcudf decimal type, but would likely require some extra conversion code as well.

Reply:
So, I whipped up this patch to `test_can_parse_avro_time_logical_types()` before I read your comment:
```
diff --git a/python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py b/python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py
index 7a275c6713..d279dff3b2 100644
--- a/python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py
+++ b/python/cudf/cudf/tests/test_avro_reader_fastavro_integration.py
@@ -404,6 +404,15 @@ def test_can_parse_avro_time_logical_types():
 
     actual = cudf_from_avro_util(schema, records)
 
+    # Make sure cudf honors our timedelta64[ms] and [us] types for millis and
+    # micros, respectively.
+    dtypes = actual.dtypes
+    assert str(dtypes['f_time_millis']) == 'timedelta64[ms]'
+    assert str(dtypes['f_time_micros']) == 'timedelta64[us]'
+
+    # N.B. We have to use timedelta64[ns] here as assert_eq() will convert both
+    #      dataframes into pandas to do the comparison, and pandas will convert
+    #      all timedelta64's into timedelta64[ns].
```

I agree that `time-millis` and `time-micros` don't have exact cudf equivalents, and that I'm shoehorning them into the duration types.  It sounds like the consensus is that I shouldn't do that.  i.e. if I make your change above, it'll mean I'll be removing the support for `time-millis` and `time-micros` as supported times (as the `TIMESTAMP_*` types definitely don't work right now, and need more attention).

I think that's fine though, the main goal was to initially add `date` support.  `time-micros` and `time-millis` seemed like easy wins, but I'll take them out as you're right, they're not semantically equivalent to any existing cudf type.

Reply:
Ok, I've ripped out the `time-millis` and `time-micros` support for now: https://github.com/rapidsai/cudf/compare/branch-23.04..cb82e821fea7a177d3e892d55a63cb70a094f86a